Child Process: stream stderr/out to a log file

danielbeardsley opened the pull request:
* Allow users of this module to specify a log file that
  should receive stdout / stderr from the webpack process
   * If it crashes, we'll actually have the output
* If the webpack proc crashes, notify the watchers
* Respond quickly if trying to ask the child a question
  when it's dead.

CC @jarstelfox

Closes #12

danielbeardsley committed:
weback-watcher: extract a function

danielbeardsley committed:
Child Processes: allow streaming the logs to a file

If the child process has some output (like a crash with
a stack trace), stream it to the log file the user
requests. Otherwise, leave it inherited (streams to parent
process stdout/err).

danielbeardsley committed:
Child Process: give feedback when child is dead

If the webpack proc dies, notify the watchers.
Also, respond quickl if trying to ask the child a question
when it's dead.

davidrans commented on webpack-watcher.js:
```diff
@@ -43,7 +44,27 @@ module.exports = (options) => {
+   function getForkOptions() {
+      const output = options.logPath ? 'pipe' : 'inherit';
+      return {
+         stdio: ['inherit', output, output, 'ipc']
```
I read the docs here, and I don't understand when/why this should be an array of 4 things (as opposed to 3, in/out/err)? https://nodejs.org/api/child_process.html#child_process_options_stdio

danielbeardsley replied:
> (for `.fork()`) When this option is provided, it overrides `silent`. If the array variant is used, it must contain exactly one item with value 'ipc' or an error will be thrown. For instance [0, 1, 2, 'ipc'].

jarstelfox commented:
CR 🌵

sctice commented on webpack-watcher.js:
```diff
@@ -38,10 +39,46 @@ module.exports = (options) => {
+         if (child.connected) {
+            child.send({event: 'isRunning'})
+         } else {
+            notifyWatchers("Webpack child process crashed, check the log");
```
Could we use `options.logPath` to point people directly at the log and _not_ suggest checking the log if `options.logPath` is falsey?

danielbeardsley replied:
Fixed.

sctice commented on webpack-watcher.js:
```diff
@@ -38,10 +39,46 @@ module.exports = (options) => {
+   function listenForExit(child) {
+      child.on('exit', function(code) {
+         console.log("Process for " + options.username + " stopped with exit code: " + code);
+         notifyWatchers("Webpack child process crashed, check the log");
```
Same note here about using `options.logPath`.

danielbeardsley replied:
Fixed.

sctice commented:
CR 🎏 but deploy_block ✋ on considering a suggestion.

danielbeardsley committed:
Webpack log: spit a line on build success / fail

Include the timestamp and the error. Now that we have a log, let's use
it.

danielbeardsley committed:
webpack log: refer to actual log file

If we know the path, let's include that in the error message.

Note: err can be an object, hence the check.

danielbeardsley commented:
un_deploy_block 👍   Good suggestions!

danielbeardsley commented:
Note we still suggest checking the log cause the output will be in the log of the parent process.

jarstelfox commented:
CR 🌵

danielbeardsley committed:
Webpack log: expose build compilation errors

These include syntax errors and other non-crashing but non-succesful
build errors.

k0rvusk0r4x commented:
Not a clue how to test this. Ideas?

Also, this is not showing up on Pulldasher but a pull is deploy blocked on it.

jarstelfox commented:
CR 🌵

danielbeardsley commented:
I'll post some screenshots of the webpack logs that we get now

Syntax error:
![image](https://user-images.githubusercontent.com/26855/70754679-abe51c80-1cec-11ea-9bb5-5b488fa72d1f.png)

Module not found:
![image](https://user-images.githubusercontent.com/26855/70754691-b7d0de80-1cec-11ea-9d27-46304c040e1c.png)

Webpack config error:
![image](https://user-images.githubusercontent.com/26855/70754756-dcc55180-1cec-11ea-983b-3deb6ea27aa9.png)

Messages when builds fail:
![image](https://user-images.githubusercontent.com/26855/70754819-0bdbc300-1ced-11ea-9412-fd2b38502798.png)

The client now sees this for many of the errors:
![image](https://user-images.githubusercontent.com/26855/70754865-2ada5500-1ced-11ea-97a1-3a65ffb7f537.png)

QA 👍 

I've also added this to pulldasher.
